File: .claude/hooks/plan-naming-enforcer.js
#!/usr/bin/env node
// PreToolUse hook: Enforce plan file naming convention
// Runs before Write operations to ensure plan files follow YYYY-MM-DD-context.md format

import fs from 'fs';
import path from 'path';
import os from 'os';

const input = JSON.parse(fs.readFileSync('/dev/stdin', 'utf8'));
const { tool_name, parameters } = input;

// Only process Write operations targeting plan files
if (tool_name !== 'Write') process.exit(0);

const filePath = parameters?.path;
if (!filePath) process.exit(0);

// Check if this is a plan file (in plans/ directory and .md extension)
const isInPlansDir = filePath.includes('/plans/') || filePath.startsWith('plans/');
const isMdFile = filePath.endsWith('.md');

if (!isInPlansDir || !isMdFile) process.exit(0);

// Extract filename from path
const filename = path.basename(filePath);

// Check if filename follows YYYY-MM-DD-context.md pattern
const datePattern = /^\d{4}-\d{2}-\d{2}-.+\.md$/;

if (datePattern.test(filename)) {
    // Already follows convention, allow it
    process.exit(0);
}

// Check if it's a random generated name that needs fixing
const randomPattern = /^[a-z]+-[a-z]+-[a-z]+\.md$/i;
if (randomPattern.test(filename)) {
    // Generate suggested name based on current date
    const today = new Date().toISOString().split('T')[0]; // YYYY-MM-DD format
    
    console.error(`[PLAN NAMING] File "${filename}" doesn't follow naming convention.`);
    console.error(`Expected format: YYYY-MM-DD-context.md`);
    console.error(`Suggestion: ${today}-your-task-description.md`);
    console.error(`Please rename the file to follow the convention in CLAUDE.md`);
    
    // Don't block the operation, just warn
    process.exit(0);
}

// For other non-conforming names, provide guidance
if (!filename.startsWith('active-context') && 
    !filename.startsWith('decisions') && 
    !filename.startsWith('progress') &&
    !filename.startsWith('session-handoff')) {
    
    const today = new Date().toISOString().split('T')[0];
    console.error(`[PLAN NAMING] Plan file "${filename}" should follow YYYY-MM-DD-context.md format.`);
    console.error(`Example: ${today}-refactor-auth-flow.md`);
}

process.exit(0);

File: .claude/hooks/plan-naming-enforcer.rs
use serde_json::Value;
use std::io::{self, Read};
use regex::Regex;
use chrono::Utc;

fn main() -> io::Result<()> {
    // Read JSON input from stdin
    let mut input = String::new();
    io::stdin().read_to_string(&mut input)?;
    
    let json: Value = serde_json::from_str(&input).unwrap_or_default();
    
    // Extract tool name and file path
    let tool_name = json["tool_name"].as_str().unwrap_or("");
    let file_path = json["parameters"]["path"].as_str().unwrap_or("");
    
    // Only process Write operations
    if tool_name != "Write" || file_path.is_empty() {
        return Ok(());
    }
    
    // Check if this is a plan file
    if !file_path.contains("/plans/") && !file_path.starts_with("plans/") {
        return Ok(());
    }
    
    if !file_path.ends_with(".md") {
        return Ok(());
    }
    
    // Extract filename
    let filename = file_path.split('/').last().unwrap_or("");
    
    // Check if filename follows YYYY-MM-DD-context.md pattern
    let date_pattern = Regex::new(r"^\d{4}-\d{2}-\d{2}-.+\.md$").unwrap();
    if date_pattern.is_match(filename) {
        return Ok(()); // Already follows convention
    }
    
    // Check if it's a random generated name
    let random_pattern = Regex::new(r"^[a-z]+-[a-z]+-[a-z]+\.md$").unwrap();
    if random_pattern.is_match(filename) {
        let today = Utc::now().format("%Y-%m-%d");
        eprintln!("[PLAN NAMING] File \"{}\" doesn't follow naming convention.", filename);
        eprintln!("Expected format: YYYY-MM-DD-context.md");
        eprintln!("Suggestion: {}-your-task-description.md", today);
        eprintln!("Please rename the file to follow the convention in CLAUDE.md");
        return Ok(());
    }
    
    // For other non-conforming names
    let system_files = ["active-context", "decisions", "progress", "session-handoff"];
    let is_system_file = system_files.iter().any(|&sf| filename.starts_with(sf));
    
    if !is_system_file {
        let today = Utc::now().format("%Y-%m-%d");
        eprintln!("[PLAN NAMING] Plan file \"{}\" should follow YYYY-MM-DD-context.md format.", filename);
        eprintln!("Example: {}-refactor-auth-flow.md", today);
    }
    
    Ok(())
}

// Cargo.toml would be:
// [package]
// name = "plan-naming-enforcer"
// version = "0.1.0"
// edition = "2021"
//
// [dependencies]
// serde_json = "1.0"
// regex = "1.0"
// chrono = { version = "0.4", features = ["serde"] }